### ISSUE TITLE
refactor-typing-protocols

### ISSUE THREAD

**Issue by MarceloCoelho1** [dpinlab/atelierflow#5]
Refactor the two base classes (Model and Metric) from ABC to a protocol. This solution is important to avoid a chain of wrappers or proxies.

The example above shows that. The Transformer class inherits the Model class, but in the MTSA framework, all models apply the proxy pattern, so in this example, we have a chain of proxies.

```python
# The transformer model in MTSA is already a proxy.
from mtsa.models.transformer import Transformer
class Transformer(Model):
  def __init__(self, **kwargs):
    self.model = Transformer(**kwargs)

  def fit(self, X):
    self.model.fit(X)

  def predict(self, X):
    return self.model.predict(X)

  def score_samples(self, X):
    return self.model.score_samples(X)
```

Using the typing protocol, we don’t need anymore that the Transformer Class inherits the Model class. Any class just needs to implement the methods defined in the Protocol to be considered a valid Model.

```python
from typing import Protocol

class Model(Protocol):
  def fit(self, X, y=None, **kwargs):
    ...
  def predict(self, X, **kwargs):
    ...

#
class Transformer():
  def __init__(self, **kwargs):
    self.model = Transformer(**kwargs)

  def fit(self, X):
    self.model.fit(X)

  def predict(self, X):
    return self.model.score_samples(X)
```

This might work in these two classes below, because the Transformer class has these methods:

```python
class TrainModelStep(Step[SplitDataResult, TrainingResult]):
  def __init__(self, model: Model):
    self.model = model

  def run(self, input_data: SplitDataResult, experiment_config: Dict[str, Any]) -> TrainingResult:
    logging.info(f"Fitting model '{self.model.__class__.__name__}'...")
    self.model.fit(input_data.X_train)
    return TrainingResult(model=self.model, X_test=input_data.X_test, y_test=input_data.y_test)

class EvaluateModelStep(Step[TrainingResult, EvaluationResult]):
  def __init__(self, metrics: List[Metric]):
    self.metrics = metrics

  def run(self, input_data: TrainingResult, experiment_config: Dict[str, Any]) -> EvaluationResult:
    logging.info("Making predictions on the test set...")
    y_pred = input_data.model.predict(input_data.X_test)
    scores = {}
    for metric in self.metrics:
      score = metric.compute(input_data.y_test, y_pred)
      scores[metric.__class__.__name__] = score
    return EvaluationResult(evaluation_scores=scores)
```

With these changes, we can integrate the code of other libraries or frameworks without inheriting the class Model. We just need to implement some methods of this class.

Resuming protocols:
If you run like a dog and bark like a dog, you are a dog without inheriting the dog :)
